### PR TITLE
Propagate io and terminal errors upward and ignore EPIPE

### LINKED ISSUE #1621: rustup panics when piping to `grep -q`
I noticed that rustup panics in the case below.

```
 RUST_BACKTRACE=1 rustup show | grep -q ^nightly
thread 'main' panicked at 'failed printing to stdout: Broken pipe (os error 32)', libstd/io/stdio.rs:700:9
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::panicking::default_hook::{{closure}}
             at libstd/sys_common/backtrace.rs:71
   2: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:227
             at libstd/panicking.rs:476
   3: std::panicking::continue_panic_fmt
             at libstd/panicking.rs:390
   4: std::panicking::begin_panic_fmt
             at libstd/panicking.rs:345
   5: std::io::stdio::_print
             at libstd/io/stdio.rs:700
             at libstd/io/stdio.rs:709
   6: rustup_init::rustup_mode::main
   7: rustup_init::run_rustup
   8: rustup_init::main
   9: std::rt::lang_start::{{closure}}
  10: main
  11: __libc_start_main
  12: <unknown>
```

### PR DESCRIPTION
**Pull request by NickeZ**
This is an alternative to: #1622

In this solution I instead ignore the EPIPE error at the highest level. I haven't contributed to rustup before, so please check the PR carefully.

Fixes #1621

### CODE REVIEW

**Comment by kinnison** [comment]
HI @NickeZ thanks for taking the time to produce this alternative patch.  I'm sorry I was away last week so didn't get a chance to review it before now.

A brief look over suggests this is a reasonable approach, but I want to take some time to consider if there're other places where we ought to handle epipe in a similar fashion (e.g. component list)

Please bear with me for a few days.

**Review by kinnison — APPROVED** [review]
I like this change. :+1:

**Comment by NickeZ** [comment]
Great, but can you check if the macro I have written is correct/idiomatic?

edit: It could probably just be a regular function...

**Comment by NickeZ** [comment]
Is it possible to add a test that runs the equivalent of `rustup show | head -1`? Could someone show me approximately where and how to do that?

File: Cargo.lock [hunk 1]
```diff
@@ -1,3 +1,5 @@
+# This file is automatically @generated by Cargo.
+# It is not intended for manual editing.
```
Comment (kinnison) [inline Cargo.lock:2]: Is this added by a new version of Cargo?

**Review by kinnison — APPROVED** [review]
Yeah, I think as a function it's a bit nicer.  I didn't mind the macro at all, but this is perhaps easier for others to grok.

Are there any other parts of the codebase where handling epipe would be useful?  I'm not asking for them in this PR, it has a :+1: already from me, but perhaps once this is in you might go for a wander and see if there's other places worth checking on?

Regarding the testing for `rustup show | head -1` I don't think we have a way to easily simulate that in our tests right now, so I don't mind not having that yet, but if you think of something I'll be pleased to review another PR (or change on this one) for it.

File: Cargo.lock (hunk 1 above)
Comment (NickeZ) [reply Cargo.lock:2]: It must have been. I only ran cargo fmt, check and build. I didn't manually change that file. do you want me to remove it?

**Comment by NickeZ** [comment]
Do you want me to squash it or anything else before merging? To me it is good to go as it is now.

**Comment by kinnison** [comment]
> Do you want me to squash it or anything else before merging? To me it is good to go as it is now.

My preference is certainly that the commits are neat and tidy, so yes, a squash would be appreciated.  

Thanks

File: Cargo.lock (hunk 1 above)
Comment (dwijnand) [reply Cargo.lock:2]: > Is this added by a new version of Cargo?

Yes: https://github.com/rust-lang/cargo/pull/6548

**Comment by NickeZ** [comment]
OK, squashed.

**Comment by nrc** [comment]
Sorry, this needs a rebase

**Comment by NickeZ** [comment]
rebased and pushed

**Comment by NickeZ** [comment]
ping

**Pull request merged**